torch/_dynamo/variables/builtin.py

```diff
@@ -1215,13 +1216,61 @@ def call_setattr(
             tx.output.side_effects.is_attribute_mutation(obj)
             and name_var.is_python_constant()
         ):
-            name = name_var.as_python_constant()
-            if name == "requires_grad" and isinstance(obj, variables.TensorVariable):
-                unimplemented(
-                    "mutating requires_grad can introduce a new leaf from non-leaf or vice versa in "
-                    "the middle of the graph, which aot_autograd does not currently know how to handle. "
-                )
-            tx.output.side_effects.store_attr(obj, name, val)
+            if isinstance(obj, variables.TensorVariable):
+                from .builder import wrap_fx_proxy
+
+                name = name_var.as_python_constant()
+                if name == "requires_grad":
+                    # TODO(voz): Make it work properly
+                    unimplemented(
+                        "mutating requires_grad can introduce a new leaf from non-leaf or vice versa in "
+                        "the middle of the graph, which aot_autograd does not currently know how to handle. "
+                    )
+                if name == "data":
+                    # Remove the old reference in tracked fakes - if we don't do this
+                    # new .data value size and shape differences will cause
+                    # tracked fakes to produce incorrect guards. This is sound because the TensorVariable
+                    # coming out of set_() below will be a new one, and get
+                    # installed in tracked fakes.
```

Comment:
Hmm I'm a little unclear on what `tracked_fakes` is used for, and why we still need this extra handling now that we have the "immutable `FakeTensorMode` infra. Just leaving the comment but I'll read the `tracked_fakes` code some more

Reply:
oh at first I was going to say `set_()` is an inplace op so it will just return the same (mutated) FakeTensor as the input".

But I see that `tracked_fakes` is a List not a Set. So I think this makes sense to me. One question I have is: does that mean that we'll end up with duplicate FakeTensors in the `tracked_fakes` list whenever we have an ordinary inplace op? And I guess that's fine because that list is only really used by `ShapEnv`, which is only using it to generate guards on the FakeTensor symbols.

Reply:
Potentially? We should test for this.